Chained.h

```diff
@@ -11,197 +13,195 @@
-        for(auto bucket : (*hashmap)){
-            for(auto kvp : bucket){
-                uint64_t index = hash(kvp.key);
-                (*newMap)[index].push_back(kvp);
-            }
-        }
-        auto temp = hashmap;
-        hashmap = newMap;
-        delete temp;
-    }
-
-    Value defaultGetReturn(){
-        std::cerr << "Invalid types for defaultGetReturn";
-        exit(1);
-    }
-
-    void printKeyValuePair(KeyValuePair kvp){
-        std::cout << " Invalid types for printing ";
-    }
-
 public:
     ChainedMap() {
-        hashmap = new std::vector<std::list<KeyValuePair>>(hashmapSize);
+        hashmap_.resize(DEFAULT_CHAINED_MAP_SIZE);
     }
 
     ChainedMap(double load) {
-        hashmap = new std::vector<std::list<KeyValuePair>>(hashmapSize);
-        MAX_LOAD_FACTOR = load;
+        hashmap_.resize(DEFAULT_CHAINED_MAP_SIZE);
+        MAX_LOAD_FACTOR_ = load;
     }
 
-    ~ChainedMap() {
-        delete hashmap;
-    }
-
-    Value get(Key k){
-        uint64_t index = hash(k);
-        for(auto kvp : (*hashmap)[index]){
-            if(k == kvp.key){
-                return kvp.value;
+    Value get(Key k) override {
+        uint64_t index = hash(k, hashmap_.size());
+        for(auto kvp : hashmap_[index]){
+            if(k == kvp.key_){
+                return kvp.value_;
             }
         }
-        return defaultGetReturn();
+        return Value{};
     }
 
-    void put(Key k, Value v) {
-        uint64_t index = hash(k);
-        for(auto kvp : (*hashmap)[index]){
-            if(kvp.key == k){
-                kvp.value = v;
+    void put(Key k, Value v) override {
+        uint64_t index = hash(k, hashmap_.size());
+        for(auto kvp : hashmap_[index]){
+            if(kvp.key_ == k){
+                kvp.value_ = v;
                 return;
             }
         }
         KeyValuePair kvp(k, v);
-        (*hashmap)[index].push_back(kvp);
-        ++numPairs;
-        if((static_cast<float>(numPairs))/(static_cast<float>(hashmapSize)) > MAX_LOAD_FACTOR){
+        hashmap_[index].push_back(kvp);
+        ++numPairs_;
+        if(getLoadFactor() > MAX_LOAD_FACTOR_){
             doubleSize();
         }
     }
 
-    void remove(Key k) {
-        uint64_t index = hash(k);
-        for(auto i = (*hashmap)[index].begin(); i != (*hashmap)[index].end(); ++i) {
-            if(k == i->key){
-                (*hashmap)[index].erase(i);
-                --numPairs;
+    void remove(Key k) override {
+        uint64_t index = hash(k, hashmap_.size());
+        for(auto i = hashmap_[index].begin(); i != hashmap_[index].end(); ++i) {
+            if(k == i->key_){
+                hashmap_[index].erase(i);
+                --numPairs_;
                 return;
             }
         }
+        std::cerr << "Chained Map does not contain the key \"" << k << "\" so it could not be removed\n";
     }
 
-    void display(bool print){
+    void display() override {
```

Comment:
I often end up writing "print the key bits of my data structure" functions.  It really helps in debugging.  And it gives me a better sense of what is important about the data and how I should think about it.

Reply:
Here is a random example.  Hmm, I should have used `const auto& tns `and `const auto& otns` (which stands for other `tns `in my mind).  `tns `stands for type name set.
```
      auto tns = tns_res.value();
      auto otns = otns_res.value();
      if (tns != otns) {
        fmt::format_to(
            std::back_inserter(buf),
            "node_entity_type_ids differ. {:4} {} {} {} {}\n", i,
            node_entity_data_[i], fmt::join(tns, ", "),
            other->node_entity_data_[i], fmt::join(otns, ", "));
        match = false;
      }
```